File: Algorithms/692. Top K Frequent Words/TopKFrequent.cs
//Author: Tushar Jaiswal
//Creation Date: 08/18/2018

/*Given a non-empty list of words, return the k most frequent elements. Your answer should be sorted by frequency from highest to lowest. If two words have the same frequency, then the word with the lower alphabetical order comes first.

Example 1:
Input: ["i", "love", "leetcode", "i", "love", "coding"], k = 2
Output: ["i", "love"]
Explanation: "i" and "love" are the two most frequent words. Note that "i" comes before "love" due to a lower alphabetical order.

Example 2:
Input: ["the", "day", "is", "sunny", "the", "the", "the", "sunny", "is", "is"], k = 4
Output: ["the", "is", "sunny", "day"]
Explanation: "the", "is", "sunny" and "day" are the four most frequent words, with the number of occurrence being 4, 3, 2 and 1 respectively.

Note: You may assume k is always valid, 1 ≤ k ≤ number of unique elements. Input words contain only lowercase letters.
Follow up: Try to solve it in O(n log k) time and O(n) extra space.*/

public class Solution {
    public IList<string> TopKFrequent(string[] words, int k) {
        SortedSet<Element> sSet = new SortedSet<Element>(new ElementComparer());
        Dictionary<string, int> map = new Dictionary<string, int>(); 
        List<string> result = new List<string>();
        foreach(string word in words)
        {
            if(map.ContainsKey(word))
            { map[word] += 1; }
            else
            { map.Add(word, 1); }
        }
        
        foreach(KeyValuePair<string, int> kvp in map)
        {
            sSet.Add(new Element(kvp.Key, kvp.Value));
            if(sSet.Count > k)
            { sSet.Remove(sSet.First()); }
        }
        
        while(sSet.Count > 0)
        { 
            result.Insert(0, sSet.First().word); 
            sSet.Remove(sSet.First());
        }
        return result;
    }
}

public class Element
{
    public string word;
    public int frequency;
    
    public Element(string word, int frequency)
    {
        this.word = word;
        this.frequency = frequency;
    }
}

public class ElementComparer : IComparer<Element>
{
    public int Compare(Element a, Element b)
    {
        if(a.frequency == b.frequency)
        { return b.word.CompareTo(a.word); }
        return a.frequency.CompareTo(b.frequency);
    }
}

File: Algorithms/347. Top K Frequent Elements/TopKFrequent.py
# Author: Tushar Jaiswal
# Creation Date: 2025-08-03

# Problem Source: https://leetcode.com/problems/top-k-frequent-elements

# Given an integer array nums and an integer k, return the k most frequent elements. You may return the answer in any order.

# Example 1:
# Input: nums = [1,1,1,2,2,3], k = 2
# Output: [1,2]

# Example 2:
# Input: nums = [1], k = 1
# Output: [1]

# Constraints:
# 1 <= nums.length <= 10^5
# -10^4 <= nums[i] <= 10^4
# k is in the range [1, the number of unique elements in the array].
# It is guaranteed that the answer is unique.

# Follow up: Your algorithm's time complexity must be better than O(n log n), where n is the array's size.

# Time complexity : O(N+Nlogk) if k<N and O(1) in the particular case of N=k. That ensures time complexity to be better than O(NlogN).
# Space complexity : O(N+k) to store the hash map with not more N elements and a heap with k elements.

import heapq

class Element:
    def __init__(self, val: int, frequency: int):
        self.val = val
        self.frequency = frequency
    
    def __lt__(self, other):
        return self.frequency < other.frequency

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        if k == len(nums):
            return nums

        frequency = Counter(nums)
        heap = []

        for val, frequency in frequency.items():
            heapq.heappush(heap, Element(val, frequency))
            if len(heap) > k:
                heapq.heappop(heap)

        return [x.val for x in heap]
